kyaml/filesys/filesystem_test.go

```diff
@@ -42,3 +52,87 @@ func testNotExistErr(t *testing.T, fs FileSystem) {
 	err = fs.Walk(path, func(_ string, _ os.FileInfo, err error) error { return err })
 	assert.Truef(t, errors.Is(err, os.ErrNotExist), "Walk should return ErrNotExist, got %v", err)
 }
+
+// setupFileSys returns file system, default directory to test in, and working directory
+func setupFileSys(t *testing.T, name string) (FileSystem, string, string) {
+	t.Helper()
+
+	switch name {
+	case disk:
+		fSys, testDir := makeTestDir(t)
+		return fSys, testDir, cleanWd(t)
+	case memoryAbs:
+		return filesysBuilders[name](), Separator, Separator
+	default:
+		require.Equal(t, memoryEmpty, name)
+		return filesysBuilders[name](), "", ""
+	}
+}
+
+func TestDemandDir(t *testing.T) {
+	for name := range filesysBuilders {
+		thisName := name
+		t.Run(thisName, func(t *testing.T) {
+			fSys, prefixPath, wd := setupFileSys(t, thisName)
+
+			d1Path := filepath.Join(prefixPath, "d1")
+			d2Path := filepath.Join(d1Path, ".d2")
+			err := fSys.MkdirAll(d2Path)
+			require.NoError(t, err)
+			require.Truef(t, fSys.Exists(d2Path), existMsg, d2Path)
+
+			tests := map[string]*struct {
+				dir      string
+				cleanDir string
+			}{
+				"Simple": {
+					d1Path,
+					d1Path,
+				},
+				"Hidden": {
+					d2Path,
+					d2Path,
+				},
+				"Relative": {
+					SelfDir,
+					wd,
+				},
+			}
+			for subName, test := range tests {
+				tCase := test
+				t.Run(subName, func(t *testing.T) {
+					cleanDir, err := DemandDir(fSys, tCase.dir)
+					require.NoError(t, err)
+					require.Equal(t, tCase.cleanDir, cleanDir.String())
+				})
+			}
+		})
+	}
+}
+
+func TestDemandDirErr(t *testing.T) {
+	for name := range filesysBuilders {
+		thisName := name
+		t.Run(thisName, func(t *testing.T) {
+			fSys, prefixPath, _ := setupFileSys(t, thisName)
+
+			fPath := filepath.Join(prefixPath, "foo")
+			err := fSys.WriteFile(fPath, []byte(`foo`))
+			require.NoError(t, err)
+			require.Truef(t, fSys.Exists(fPath), existMsg, fPath)
+
+			tests := map[string]string{
+				"Empty":        "",
+				"File":         fPath,
+				"Non-existent": filepath.Join(prefixPath, "bar"),
+			}
+			for subName, invalidPath := range tests {
+				invalid := invalidPath
```

Comment:
Admittedly, I don't fully understand why in tests we need to capture range variables, but I don't think it's necessary here?

Reply:
So we need to capture the variable value during the iteration because range variables are alive during the entire iterative process, and so its value will keep changing. `t.Run()` calls the function we define and pass in asynchronously, so by the actual sub-test runs, the range variable might be on a completely different value than that of the corresponding iteration.

We do need `invalid` in this case because we use it in our sub-test code. Whether or not I change its name from `invalidPath` to `invalid` is a different matter and I will apply your judgment from `thisName`.

Reply:
SG, thanks for the explanation!